### PR TITLE
feat(user): Report profiles

### LINKED ISSUE #11891: Reporting Profiles.
Refers a simple email based reporting. https://github.com/FreeCodeCamp/FreeCodeCamp/issues/10807#issuecomment-264329567

### PR DESCRIPTION
**Pull request by raisedadead**
This adds a simple email-based mechanism to report profiles for abuse. An email with text from the report is sent to Free Code Camp's support account with the reporter's account in copy. This also adds the reporter's details to the report for follow ups.

Tested locally.

Closes #11891

### CODE REVIEW

File: server/boot/user.js
Comment (BerkeleyTrue) [inline server/boot/user.js]: This should be a middleware and both the get request and the api should block those that have not been authenticated.

File: server/boot/user.js
Comment (BerkeleyTrue) [inline server/boot/user.js]: Why are you sending the users email to the reporter?

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Comment by BerkeleyTrue** [comment]
This seems something that can easily be abused and does not really solve the problem. 

I'm not sure just sending an email to quincy is the right thing to do. This could potentially make our inbox unmanageable.

@QuincyLarson What are you thoughts?

File: server/boot/user.js
Comment (raisedadead) [inline server/boot/user.js]: We are showing a notif to the reporter. This is the reporter's email.

We are keeping them in the copy to continue the follow-up, plus having them in copy means they won't be spamming with multiple reports.

**Comment by raisedadead** [comment]
@BerkeleyTrue thanks for the QA.

Yeah, it's a very simple, light-weight, mechanism, as per @QuincyLarson's [comment](https://github.com/FreeCodeCamp/FreeCodeCamp/issues/10807#issuecomment-264329567), we aren't looking to hit the database.

Regarding safeguards against abuse, we are aiming to have this towards logged in users with verified emails only. Each report is sent as a copy to the reporter. This aims to reduce the chances of spam mailing.

But of course, we should consider this least on the list of our priorities.

I will fix the routing issues in your review and update the PR.

**Comment by raisedadead** [comment]
On that note, maybe I'll add some recorded GIF's to showcase the flow, if that would help.

**Comment by BerkeleyTrue** [comment]
Alright, let's give it a try and see how it works out.

**Review by BerkeleyTrue — APPROVED** [review]

**Comment by raisedadead** [comment]
Please wait for fixes to Berkeley's earlier review comments before merging, will update when ready.

**Comment by raisedadead** [comment]
Ready for QA. Have addressed the earlier comments and added the unAuth and the non-verified email middlewares.

File: server/boot/user.js
Comment (BerkeleyTrue) [inline server/boot/user.js]: sending status code should be done when the endpoint is an api. 

When the endpoint is sending back to a page the redirect should also send them to a page.

So here we want to redirect unauthenticated(not logged in) users to either the `map` or `login`

File: server/utils/middleware.js
Comment (BerkeleyTrue) [inline server/utils/middleware.js]: We can shorten this to just `ifNotVerifiedRedirectToSettings` as we will be using the verified field more often now we don't need the extra `email` qualifier.

File: server/boot/user.js
Comment (BerkeleyTrue) [inline server/boot/user.js]: API endpoints should always include the `next` argument.

File: server/boot/user.js
Comment (BerkeleyTrue) [inline server/boot/user.js]: errors should be sent to the default express handler using `return next(err)`. This prevents having to write duplicate code.

You can still define the redirect by adding the `redirectTo` property to the error instance.

https://github.com/FreeCodeCamp/FreeCodeCamp/blob/staging/server/middlewares/error-handlers.js#L32

File: server/boot/user.js
Comment (BerkeleyTrue) [inline server/boot/user.js]: req.params.username could be `undefined` causing this line to crash. 

Also, this line and the two below can be done declaratively using object destructuring.

File: server/boot/user.js
Comment (BerkeleyTrue) [inline server/boot/user.js]: Both the `reportDescription` and username should be sanitized just in case. We don't want malicious scripts to be sent to out email

File: server/boot/user.js
Comment (BerkeleyTrue) [inline server/boot/user.js]: This is redirect the reporting user to their own user profile. Is this intentional? shouldn't this send them back to the page they were on using req.params.username?

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

File: server/boot/user.js
Comment (raisedadead) [inline server/boot/user.js]: Fixed.

**Comment by raisedadead** [comment]
This is ready for QA. I have added a custom sanitizer (source is quoted in the comments).

**Comment by BerkeleyTrue** [comment]
Why are you adding a custom sanitizer? Why not use the one built into express-validator?

**Comment by raisedadead** [comment]
Going by the docs, they have just basic escaping as default.
see: [express-validator#sanitizer](https://github.com/ctavan/express-validator#sanitizer)

More sanitizers are available from, their dependency in [chriso/validator.js](https://github.com/chriso/validator.js), but they apparently removed support for scripts, more details in their readme.

I went by a SO answer http://stackoverflow.com/a/430240/1932901, it was aimed for client side cleanup, by solves our purposes, for API side validation.

I tested, it recursively removes all sorts of tags and dirty elements from the text. 
Again, very basic but does a much cleaner job.

I am open to suggestions though.

**Comment by BerkeleyTrue** [comment]
ok sounds good.

File: server/middlewares/validator.js
Comment (BerkeleyTrue) [inline server/middlewares/validator.js]: use `let` instead of `var`

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Comment by raisedadead** [comment]
Ready for a QA.

**Review by BerkeleyTrue — APPROVED** [review]

**Comment by BerkeleyTrue** [comment]
code lgtm

**Comment by QuincyLarson** [comment]
@raisedadead This works amazingly. Thank you for implementing this. :shipit:

**Comment by raisedadead** [comment]
Happy to help!

**Pull request merged**
